src/Common/src/CoreLib/System/IO/PathInternal.cs

```diff
@@ -117,24 +126,61 @@ internal static bool AreRootsEqual(string first, string second, StringComparison
         internal static string RemoveRelativeSegments(string path, int rootLength)
         {
             Debug.Assert(rootLength > 0);
-            bool flippedSeparator = false;
 
+            Span<char> initialBuffer = stackalloc char[MaxStackPath];
+            ValueStringBuilder sb = new ValueStringBuilder(initialBuffer);
+
+            bool pathNotChanged = RemoveRelativeSegmentsHelper(path.AsSpan(), rootLength, ref sb);
+
+            // If we haven't changed the source path, return the original.
+            // RemoveRelativeSegmentsHelper can only modifiy the last directory separator in the root segment.
+            // It happens when only root segment is left after removing the relative segments from the path.
+            // eg RemoveRelativeSegmentsHelper will return C: in sb in cases like C:\.\, C:\..\ So in such cases we just return the root segment from the original path.
+            return pathNotChanged ? path :
+                sb.Length >= rootLength ? sb.ToString() :
+                path.Substring(0, rootLength);
+        }
+
+        /// <summary>
+        /// Try to remove relative segments from the given path (without combining with a root).
+        /// </summary>
+        /// <param name="rootLength">The length of the root of the given path</param>
+        internal static void RemoveRelativeSegments(ReadOnlySpan<char> path, int rootLength, ref ValueStringBuilder sb)
+        {
+            Debug.Assert(rootLength > 0);
+            RemoveRelativeSegmentsHelper(path, rootLength, ref sb);
+            
+            // RemoveRelativeSegmentsHelper can only modifiy the last directory separator in the root segment.
+            // It happens when only root segment is left after removing the relative segments from the path.
+            // eg RemoveRelativeSegmentsHelper will return C: in sb in cases like C:\.\, C:\..\
+            // We restore the directory separator in such cases. The sb will now contain the root segment as value.
+            if (sb.Length < rootLength)
```

Comment:
This seems like a bug to me. Can't we modify RemoveRelativeSegmentsHelper to not ever eat into the rootLength? At first glance it seems that we're fixing up a flaw in the helper.